plugins/catalog-react/src/alpha/predicates/createEntityPredicateSchema.ts

```diff
@@ -14,33 +14,44 @@
  * limitations under the License.
  */
 
-import { EntityPredicate, EntityPredicateValue } from './types';
+import {
+  EntityPredicate,
+  EntityPredicateExpression,
+  EntityPredicatePrimitive,
+  EntityPredicateValue,
+} from './types';
 import type { z as zImpl, ZodType } from 'zod';
 
 /** @internal */
 export function createEntityPredicateSchema(z: typeof zImpl) {
-  const primitiveSchema = z.union([z.string(), z.number(), z.boolean()]);
-
-  const comparableValueSchema = z.union([
-    primitiveSchema,
-    z.array(primitiveSchema),
-  ]);
+  const primitiveSchema = z.union([
+    z.string(),
+    z.number(),
+    z.boolean(),
+  ]) as ZodType<EntityPredicatePrimitive>;
 
   // eslint-disable-next-line prefer-const
   let valuePredicateSchema: ZodType<EntityPredicateValue>;
 
+  const expressionSchema = z.lazy(() =>
+    z.union([
+      z.record(z.string().regex(/^(?!\$).*$/), valuePredicateSchema),
+      z.record(z.string().regex(/(?!\$)+/), z.never()),
```

Comment:
```suggestion
      z.record(z.string().regex(/^\$/), z.never()),
```

you meant something like this, right? The given regex does not match strings that start with `$`

Reply:
You are right. Repaired.